src/python/pants/backend/go/util_rules/build_pkg_target.py

```diff
@@ -214,25 +257,81 @@ async def setup_build_go_package_target_request(
             "message!"
         )
 
-    all_deps = await Get(Targets, DependenciesRequest(target[Dependencies]))
-    maybe_direct_dependencies = await MultiGet(
-        Get(FallibleBuildGoPackageRequest, BuildGoPackageTargetRequest(tgt.address))
-        for tgt in all_deps
-        if (
-            tgt.has_field(GoPackageSourcesField)
-            or tgt.has_field(GoThirdPartyPackageDependenciesField)
-            or bool(maybe_get_codegen_request_type(tgt, union_membership))
+    all_direct_dependencies = await Get(Targets, DependenciesRequest(target[Dependencies]))
+
+    first_party_dep_import_path_targets = []
+    third_party_dep_import_path_targets = []
+    codegen_dep_import_path_targets = []
+    for dep in all_direct_dependencies:
+        if dep.has_field(GoPackageSourcesField):
+            first_party_dep_import_path_targets.append(dep)
+        elif dep.has_field(GoThirdPartyPackageDependenciesField):
+            third_party_dep_import_path_targets.append(dep)
+        elif bool(maybe_get_codegen_request_type(dep, union_membership)):
+            codegen_dep_import_path_targets.append(dep)
+
+    first_party_dep_import_path_results = await MultiGet(
+        Get(FirstPartyPkgImportPath, FirstPartyPkgImportPathRequest(tgt.address))
+        for tgt in first_party_dep_import_path_targets
+    )
+    first_party_dep_import_paths = {
+        result.import_path: tgt.address
+        for tgt, result in zip(
+            first_party_dep_import_path_targets, first_party_dep_import_path_results
+        )
+    }
+
+    pkg_dependency_addresses = []
+    for dep_import_path, address in first_party_dep_import_paths.items():
+        if dep_import_path in imports:
+            pkg_dependency_addresses.append(address)
+    for dep_tgt in third_party_dep_import_path_targets:
+        dep_import_path = dep_tgt[GoImportPathField].value
+        if dep_import_path in imports:
+            pkg_dependency_addresses.append(dep_tgt.address)
```

Comment:
`pkg_dependency_addresses.update(dep_tgt.address for dep_tgt in third_party_dep_import_path_targets if dep_tgt[GoImportPathField].value in imports)`